### PR TITLE
Model configuration for PyFunc flavor

### PR DESCRIPTION
**Pull request by santiagxf**
## Related Issues/PRs

None.

## What changes are proposed in this pull request?

We are introducing the concept of model load configuration, as an experimental feature. Model load configuration gives model builders the capability to provide parameters at model's loading time for the pyfunc flavor. Such configuration is available during model loading procedure and can also be overridden by the caller of the function `mlflow.pyfunc.load_model`.

**How this functionality related with `params` recently introduced?**

You can think of `params` as an inference configuration at **predict** time that is available to the model consumer. The parameters are visible to the model **consumer** through the signature. Think of `model_config` as model configuration at **load** time. The parameters are visible to the model **publisher** through the model's medatada and can't be accessed by the model consumer. This provides a mechanism to protect the builder from end users changing the parameters of the model but at the same time enable the model operator (who is deploying it) to change how the model is executed.

## How is this patch tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [x] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/utils/model_utils.py [hunk 1]
```diff
@@ -191,3 +194,70 @@ def _validate_onnx_session_options(onnx_session_options):
                 raise ValueError(
                     f"Value for key {key} in onnx_session_options should be >= 0, not {value}"
                 )
+
+
+def _update_inference_config(
```
Comment (WeichenXu123) [inline mlflow/utils/model_utils.py:199]:
```suggestion
def _get_overridden_inference_config(
```

File: mlflow/utils/model_utils.py [hunk 2]
```diff
@@ -191,3 +194,70 @@ def _validate_onnx_session_options(onnx_session_options):
...
+    overrides = {}
+    if env_overrides := os.getenv("MLFLOW_PYFUNC_INFERENCE_CONFIG"):
+        logger.debug(
+            "Inference configuration is being loaded from ``MLFLOW_PYFUNC_INFERENCE_CONFIG``"
+            " environ."
+        )
+        overrides.update(dict(json.loads(env_overrides)))
+
+    if load_config:
+        overrides.update(load_config)
```
Comment (WeichenXu123) [inline mlflow/utils/model_utils.py:218]: Q:

What if `MLFLOW_PYFUNC_INFERENCE_CONFIG` and load_config both exists for the same param key ?
shall we forbid this case or tell user which one will be overridden by another one ?

File: mlflow/utils/model_utils.py [hunk 3]
```diff
@@ -191,3 +194,70 @@ def _validate_onnx_session_options(onnx_session_options):
...
+    if len(allowed_config) < len(overrides):
+        ignored_keys = set(overrides.keys()) - set(allowed_config.keys())
+        logger.warning(
+            f"Argument(s) {', '.join(ignored_keys)} were ignored since they are not"
+            " section of the ``pyfunc`` flavor. Use ``inference_config`` when logging the model"
+            " to allow inference configuration. Allowed configuration includes"
+            f" {', '.join(pyfunc_config.keys())}"
+        )
+
+    return pyfunc_config.update(allowed_config)
```
Comment (WeichenXu123) [inline mlflow/utils/model_utils.py:243]:
```suggestion
    pyfunc_config.update(allowed_config)
    return pyfunc_config
```

Note that `dict.update(...)` returns None

File: mlflow/utils/model_utils.py [hunk 4]
```diff
@@ -191,3 +194,70 @@ def _validate_onnx_session_options(onnx_session_options):
...
+    """
+
+    def is_jsonable(value):
+        try:
+            json.dumps(value)
+            return True
+        except (TypeError, OverflowError):
+            return False
+
+    if inference_config and not is_jsonable(inference_config):
```
Comment (WeichenXu123) [inline mlflow/utils/model_utils.py:259]: We should verify `inference_config` is a dict firstly if it is not None.

File: mlflow/utils/model_utils.py [hunk 5]
```diff
@@ -191,3 +194,70 @@ def _validate_onnx_session_options(onnx_session_options):
...
+            f"Argument(s) {', '.join(ignored_keys)} were ignored since they are not"
+            " section of the ``pyfunc`` flavor. Use ``inference_config`` when logging the model"
+            " to allow inference configuration. Allowed configuration includes"
+            f" {', '.join(pyfunc_config.keys())}"
+        )
+
+    return pyfunc_config.update(allowed_config)
+
+
+def _validate_inference_config(inference_config):
```
Comment (WeichenXu123) [inline mlflow/utils/model_utils.py:246]: I suggest that we should also add type validation for inference_config values, similar to inference params

File: mlflow/utils/model_utils.py (hunk 2 above)
Comment (santiagxf) [reply mlflow/utils/model_utils.py:218]: Good question. In the original design doc we stablished a priority and overriding mechanism that follows the following priority:
1. Values indicated in the parameter `inference_config` of `load_model`.
2. Values indicated in the environ `MLFLOW_PYFUNC_INFERENCE_CONFIG`.
3. Values indicated in the model metadata when using `log_model` or `save_model`.

However, I see your point of this maybe being complicated. We can allow overriding from 1) to 3) while throw and exception when 1) and 2) are both indicated. We can also keep this priority and start warning when overriding is happening. What do you think?

File: mlflow/utils/model_utils.py (hunk 5 above)
Comment (santiagxf) [reply mlflow/utils/model_utils.py:246]: We touch on this point in the last review of the feature. The problem with typing is that there are several cases where users may need complex typing. Typing would become very quickly a bit complex. In all the case we reviewed with users though, it was about a nested dictionary which is why we are placing the restriction of JSON-serializable.  Parameters typing was very important because you need to transfer data over the wire and hence typing may be tricky. Also, it's exposed to the consumer so you probably want to generate swaggers out of it so the consumer knows what can be indicated. However, in this scenario, `inference_config` is not accessible to the final consumer. It's accessible to the model provider and the person deploying the model. We are placing the assumption that implicit typing will be enough here. In the same case we use environment variables. Have you seen any issue with that approach on environment variables maybe that can inform if this is a good or a bad call?

File: mlflow/utils/model_utils.py (hunk 4 above)
Comment (santiagxf) [reply mlflow/utils/model_utils.py:260]: Great suggestion!

File: docs/source/models.rst [hunk 6]
```diff
@@ -914,6 +914,32 @@ Furthermore, if you want to run model inference in the same environment used in
...
+
+Alternatively, you can load the PyFunc model and inspect the `inference_config` property:
+
+.. code-block:: python
+
+    pyfunc_model = mlflow.pyfunc.load_model(model_uri)
+    pyfunc_model.inference_config
+
+Inference configuration can be changed at loading time by indicating `inference_config` parameter in the 
+:py:func:`mlflow.pyfunc.load_model` method:
```
Comment (harupy) [inline docs/source/models.rst:933]: Can we rename `inference_config` to `load_config` or `load_params`?

Comment (santiagxf) [reply docs/source/models.rst:933]: I the original proposal we used the name `params` which is close to `load_params`. However, we thought that would be confusing with the `params` at predict time. I completely see your point cause this configuration may be used for inference but more generally to load the model. If I have to vote, I would rather use `config` but `params` also works. I would remove the `load` word because it will be redundant when you read it at `mlflow.pyfunc.load_model(load_config=...)`. Which one do you think conveys better the idea?

Comment (santiagxf) [reply docs/source/models.rst]: I changing it to `model_config`. I think it plays well with `model_uri`... `model_config` while we don't use the same `params` word we use for predict.

File: mlflow/utils/model_utils.py [hunk 7]
```diff
@@ -191,3 +194,78 @@ def _validate_onnx_session_options(onnx_session_options):
...
+        logger.warning(
+            f"Argument(s) {', '.join(ignored_keys)} were ignored since they are not valid keys in"
+            " the corresponding section of the ``pyfunc`` flavor. Use ``model_config`` when"
+            " logging the model to include the keys you plan to indicate. Current allowed"
+            " configuration includes"
+            f" {', '.join(pyfunc_config.keys())}"
+        )
+
+    pyfunc_config.update(allowed_config)
+    return pyfunc_config
```
Comment (BenWilson2) [inline mlflow/utils/model_utils.py:245]: Could we do set operations instead of dict length comparisons to be more exact on the keys being overridden?

```suggestion
  valid_keys = set(pyfunc_config.keys()) & set(overrides.keys())
  ignored_keys = set(overrides.keys()) - valid_keys
  allowed_config = {key: overrides[key] for key in valid_keys}
  if ignored_keys:
      logger.warning(
          f"Argument(s) {', '.join(ignored_keys)} were ignored since they are not valid keys in"
          " the corresponding section of the ``pyfunc`` flavor. Use ``model_config`` when"
          " logging the model to include the keys you plan to indicate. Current allowed"
          f" configuration includes {', '.join(pyfunc_config.keys())}"
      )
  pyfunc_config.update(allowed_config)
  return pyfunc_config
```

File: mlflow/utils/model_utils.py [hunk 8]
```diff
@@ -191,3 +194,78 @@ def _validate_onnx_session_options(onnx_session_options):
...
+        if isinstance(model_config, dict) and all(
+            isinstance(key, str) for key in model_config.keys()
+        ):
+            if not is_jsonable(model_config):
+                raise MlflowException(
+                    "Some of the values indicated in ``model_config`` are not "
+                    "supported. Only JSON-serializable data types can be indicated."
+                )
+        else:
+            raise MlflowException("``model_config`` has to be of type ``dict`` with string keys.")
```
Comment (BenWilson2) [inline mlflow/utils/model_utils.py:271]: Could we short-circuit the return and handle the two separate errors that might arise:
1. A dict with non-string keys was passed (invalid model_config structure) 
2. A dict with string keys and non-encodable objects was passed

```suggestion
def _validate_pyfunc_model_config(model_config):
    """
    Validates the values passes in the model_config section. There are no typing
    restrictions but we require them being JSON-serializable.
    """

    if not model_config:
        return

    if not isinstance(model_config, dict) or not all(isinstance(key, str) for key in model_config):
        raise MlflowException("An invalid ``model_config`` structure was passed. ``model_config`` must be of type ``dict`` with string keys.")
    
    try:
        json.dumps(model_config)
    except(TypeError, OverflowError):
        raise MlflowException("Values in the provided ``model_config`` are of an unsupported type. Only JSON-serializable data types can be provided as values.")
```

Or something similar to that?

File: docs/source/models.rst [hunk 9]
```diff
@@ -2943,26 +2977,34 @@ For `transformers` inference, there are two ways to pass in additional arguments
         generate_signature_output(model, data),
     )
 
-    # Define an inference_config
-    inference_config = {
+    # Define an model_config
+    model_config = {
```
Comment (BenWilson2) [inline docs/source/models.rst:2981]: can we rename these references to ``load_config`` please?

Comment (santiagxf) [reply docs/source/models.rst:2981]: @BenWilson2 are you suggesting to name the concept `load_config` instead of `model_config`? I have mixed feelings about that. Specially consider the syntax when we say `mlflow.pyfunc.log_model(...., load_config=..)`. Indicating how to load in an instruction that saves things doesn't look good to me. Any particular reason `model_config` doesn't look appealing?

File: mlflow/pyfunc/__init__.py [hunk 10]
```diff
@@ -357,6 +372,8 @@ def add_to_model(
             params[ENV][EnvType.CONDA] = conda_env
         if python_env:
             params[ENV][EnvType.VIRTUALENV] = python_env
+    if model_config:
+        params[MODEL_CONFIG] = model_config
     return model.add_flavor(FLAVOR_NAME, **params)
```
Comment (serena-ruan) [inline mlflow/pyfunc/__init__.py:377]: Instead of saving load_config in params, could we add an attribute for Model called `load_config` directly, and save load_config when saving/logging the model? e.g. https://docs.google.com/document/d/11sRxfUIXsdsQkE_7bZTviK0fFyhP9PNIr2dqrIT05ew/edit#heading=h.ggqwwxf0svm

File: mlflow/pyfunc/__init__.py [hunk 11]
```diff
@@ -634,8 +663,15 @@ def load_model(
...
     _add_code_from_conf_to_system_path(local_path, conf, code_key=CODE)
     data_path = os.path.join(local_path, conf[DATA]) if (DATA in conf) else local_path
+    model_config = _get_overridden_pyfunc_model_config(
+        conf.get(MODEL_CONFIG, None), model_config, _logger
+    )
+
     try:
-        model_impl = importlib.import_module(conf[MAIN])._load_pyfunc(data_path)
+        if model_config:
+            model_impl = importlib.import_module(conf[MAIN])._load_pyfunc(data_path, model_config)
```
Comment (serena-ruan) [inline mlflow/pyfunc/__init__.py:672]: This won't work for modules that '_load_pyfunc' doesn't contain 'model_config'. 
Instead, could we used **kwargs in def load_model? And kwargs passed here should override existing load_config saved with the Model.

File: mlflow/pyfunc/__init__.py [hunk 12]
```diff
@@ -691,12 +727,14 @@ def _load_model_or_server(model_uri: str, env_manager: str):
...
     :param env_manager: The environment manager to load the model.
+    :param model_config: The model configuration to use by the model, only if the model
+                             accepts it.
     :return: A _ServedPyFuncModel for non-local ``env_manager``s or a PyFuncModel otherwise.
     """
     from mlflow.pyfunc.scoring_server.client import ScoringServerClient, StdinScoringServerClient
 
     if env_manager == _EnvManager.LOCAL:
-        return load_model(model_uri)
+        return load_model(model_uri, model_config=model_config)
```
Comment (serena-ruan) [inline mlflow/pyfunc/__init__.py:737]: With above update, we can read load_config directly from the model's file, then no need to update this.

File: mlflow/pyfunc/model.py [hunk 13]
```diff
@@ -332,7 +353,7 @@ def _save_model_with_class_artifacts_params(
     _PythonEnv.current().to_yaml(os.path.join(path, _PYTHON_ENV_FILE_NAME))
 
 
-def _load_pyfunc(model_path):
+def _load_pyfunc(model_path: str, model_config: Dict[str, Any] = None):
```
Comment (serena-ruan) [inline mlflow/pyfunc/model.py:356]: Same here, we should use **kwargs here to support extra configs passed in during load_model.
We should read load_config from the model file first, then update with additional passed-in configs.

File: mlflow/transformers.py [hunk 14]
```diff
@@ -217,7 +217,7 @@ def save_model(
     processor=None,
     task: Optional[str] = None,
     model_card=None,
-    inference_config: Optional[Dict[str, Any]] = None,
+    model_config: Optional[Dict[str, Any]] = None,
```
Comment (serena-ruan) [inline mlflow/transformers.py:220]: As discussed, could we revert this change and keep inference_config for now?
We should add extra load_config that's consistent with pyfunc's save_model/log_model.

File: tests/transformers/test_transformers_model_export.py [hunk 15]
```diff
@@ -3640,7 +3640,7 @@ def test_whisper_model_supports_timestamps(raw_audio_file, whisper_pipeline):
         model_info = mlflow.transformers.log_model(
             transformers_model=whisper_pipeline,
             artifact_path="model",
-            inference_config=inference_config,
+            model_config=model_config,
```
Comment (serena-ruan) [inline tests/transformers/test_transformers_model_export.py:3643]: Let's revert changes in this file, and add other tests for 
1. saving/logging model with load_config works fine
2. pass extra kwargs in mlflow.pyfunc.load_model will update the configs saved, and applied during loading

For model serving and other flavors support, I can update in follow-up PRs after we get this one  merged in :)

**Review by serena-ruan — COMMENTED** [review]
Thanks @santiagxf ! As we discussed yesterday, could you rename `model_config` to `load_config`? And I left several comments, basically I prefer saving `load_config` as a field in `Model`, and add `kwargs` in mlflow.pyfunc.load_model to support passing extra configs for loading.

File: mlflow/pyfunc/__init__.py (hunk 11 above)
Comment (santiagxf) [reply mlflow/pyfunc/__init__.py:672]: It will. Take a closer look at the `if` statement. The change is backward compatible as models without a `load_config` always load the older version of the signature. There is an assumption though that users using `load_config` when saving a model will also provide code that can handle it. Shall we add a validation to it? I don't think we do that today for existing requirements. (for instance, to check that there is a method `predict`, the method has `input` parameters, etc.). Regarding using kwargs... I would suggest to do not. That can mean trouble in the future if we plan to add new parameters that start to conflict with those from the user. I rather keep `load_config`.

File: mlflow/pyfunc/__init__.py (hunk 10 above)
Comment (santiagxf) [reply mlflow/pyfunc/__init__.py:377]: As we discussed yesterdat, the proposal considers model configuration for the `pyfunc` flavor. Adding it to the root of model will imply this is something available to all the flavors, which is not the case because it will force all the flavors to follow the same pattern, which is an unnecessary enforcement. Does it make sense?

File: mlflow/pyfunc/__init__.py (hunk 12 above)
Comment (santiagxf) [reply mlflow/pyfunc/__init__.py:737]: It is required. Take a look at how `mlflow.evaluate()` works.

File: mlflow/pyfunc/model.py (hunk 13 above)
Comment (santiagxf) [reply mlflow/pyfunc/model.py:356]: Same comments as above, kwargs is a dangerous implementation for future compatibility. It was discussed in the [original doc](https://docs.google.com/document/d/1sLR6SCby0hLRLSHonPbnzLpS-7_J4Clk/edit?usp=sharing&ouid=106956464969714561112&rtpof=true&sd=true) we shared.

File: mlflow/pyfunc/__init__.py (hunk 10 above)
Comment (serena-ruan) [reply mlflow/pyfunc/__init__.py:377]: We should only add `load_config` to flavors we would like to support. For example, in sklearn flavor we won't add `load_config` in its save_model or log_model function, then the attribute will just be None for those models.

Comment (BenWilson2) [reply mlflow/pyfunc/__init__.py:377]: +1 to @serena-ruan 's suggestion. I don't think that it'll be an all or nothing sort of deal. There are flavors that don't have the concept of providing these on-load parameters, which is fine. Provided that the logic is applied such that flavors that are not supportive of any overrides to loading configuration aren't required to have the user define anything (and not to expose an arg on load for this behavior that would effectively be a no-op), it will help to enable more use cases that are not strictly related to model serving / using a pyfunc representation of the model. Being able to override for static batch processing is likely to be a worthwhile feature.

File: mlflow/transformers.py (hunk 14 above)
Comment (BenWilson2) [reply mlflow/transformers.py:220]: There are some dependencies on our side with `inference_config` utilization with our inference platform. We should avoid modifying this for `transformers` :) Maybe we'll make that breaking change with 3.0

File: mlflow/pyfunc/__init__.py (hunk 10 above)
Comment (serena-ruan) [reply mlflow/pyfunc/__init__.py:377]: @santiagxf I think you could take another look at my design doc, and we could discuss on the design doc first if there're anything you would like to change :D Then we can implement according to the design doc.

File: mlflow/transformers.py (hunk 14 above)
Comment (santiagxf) [reply mlflow/transformers.py:220]: @BenWilson2, any dependency on existing platforms should be transparently handled because the change is backward compatible. If you want users to continue using the old way we can mark it as deprecated instead of removing from signature and doing the breaking change in 3.0. But I don't think we should keep both. It will be just a confusing signature for users with two options to achieve the same. What do you think?

File: mlflow/pyfunc/__init__.py (hunk 10 above)
Comment (santiagxf) [reply mlflow/pyfunc/__init__.py:377]: @serena-ruan, let me suggest doing it the other way around. This PR follows the original doc that we circulated for the feature so it doesn't make sense to have a discussion over a document that describes different approaches. What do you think about that?

Comment (serena-ruan) [reply mlflow/pyfunc/__init__.py:377]: The initial doc was for params and load_configs that are mixed together. That doc I wrote is for load_config specifically, now that we already have params for inference, we should follow the new doc ( and we already had internal discussion on the design doc).

File: mlflow/transformers.py (hunk 14 above)
Comment (dbczumar) [reply mlflow/transformers.py:220]: yes, let's deprecate `inference_config` and introduce `model_config`. When users specify `inference_config`, we can print a warning. @BenWilson2 @serena-ruan can you communicate this to the Databricks Model Serving team?

File: docs/source/models.rst (hunk 9 above)
Comment (dbczumar) [reply docs/source/models.rst:2995]: `model_config` is fine

File: mlflow/pyfunc/__init__.py (hunk 11 above)
Comment (dbczumar) [reply mlflow/pyfunc/__init__.py:672]: I think @serena-ruan 's point is that, if I pass `model_config` to `mlflow.pyfunc.load_model()` for a model that doesn't support `model_config`, I get a nasty "TypeError: _load_pyfunc() takes 1 positional argument but 2 were given`

We should handle that case, but I don't think `**kwargs` helps / is the right way to do it. Can we do an `inspect.signature()`?

File: mlflow/pyfunc/__init__.py (hunk 10 above)
Comment (dbczumar) [reply mlflow/pyfunc/__init__.py:377]: @serena-ruan I have some concerns about the new doc's use of `**kwargs`. We shouldn't use **kwargs for load-time config passed to `mlflow.pyfunc.load_pyfunc()` because it's very likely to lead to future compatibilities. E.g. say we want to introduce a new argument to mlflow.pyfunc.load_pyfunc() in the future called `environment` that lets users load the model in a virtualenv / conda env; we wouldn't be able to safely add that arg because someone might have a model config called `environment`. **kwargs are not future-proof; I prefer @santiagxf 's implementation here.

Regarding adding `model_config` to the pyfunc flavor or to the top-level model config, I'm again aligned with @santiagxf: `params[MODEL_CONFIG]` makes sense. If we put the model config at the top level, it means that the flavor-specific `load_model()` implementations would all have to support `model_config`. This would create confusion with existing APIs like `mlflow.pytorch.load_model()` which accept arguments like `device`: should users pass `device` or `model_config`? IMO, flavor-specific `load_model()` APIs should take **kwargs or enumerate the framework-specific load-time args that they support, and pyfunc should use a more generic `model_config` dict, which is exactly what @santiagxf's PR does.

Can we move forward with the current PR as-is?

File: mlflow/pyfunc/__init__.py (hunk 11 above)
Comment (dbczumar) [reply mlflow/pyfunc/__init__.py:672]: This is handled correctly by the behavior of `_get_overridden_pyfunc_model_config` and the  `if` check. @santiagxf can we add a test for this case? Otherwise, I think the implementation is great

**Review by dbczumar — APPROVED** [review]
LGTM! Thanks @santiagxf ! As discussed in some of the inline comments, I think this PR takes a very reasonable / correct approach and would prefer to move forward with it & merge ASAP. @BenWilson2 @serena-ruan  let me know if there are concerns I'm missing, but this seems wonderful

File: mlflow/transformers.py (hunk 14 above)
Comment (BenWilson2) [reply mlflow/transformers.py:220]: Will do!

**Review by BenWilson2 — APPROVED** [review]
Sorry for the delay (been out this last week) LGTM! Merging :)

**Comment by BenWilson2** [comment]
Hi @santiagxf looks like there's a merge conflict. Do you want me to resolve it and merge tomorrow, or do you want to handle it?

**Comment by santiagxf** [comment]
> Hi @santiagxf looks like there's a merge conflict. Do you want me to resolve it and merge tomorrow, or do you want to handle it?

I can take care of it today!

File: mlflow/pyfunc/__init__.py (hunk 11 above)
Comment (santiagxf) [reply mlflow/pyfunc/__init__.py:675]: Just added a test case for the scenario @dbczumar!
